tests/test_api.py

```diff
@@ -786,3 +786,160 @@ def test_product_components(client, api_path):
     response = client.get(f"{api_path}/components?products={rhel_br.name}")
     assert response.status_code == 200
     assert response.json()["count"] == 1
+
+
+@pytest.mark.django_db(databases=("default", "read_only"), transaction=True)
+def test_component_provides_sources_upstreams(client, api_path):
+    """
+    Taking the following component node structure:
+    /
+    └──(SOURCE) root_comp: top level root component
+        ├──(SOURCE) upstream_comp
+        └──(PROVIDES) dep1_comp
+
+    where upstream is defined as a child node of a root comp and dep1 is also defined as a root component
+    /
+    └──(SOURCE) dep1_comp: top level root component
+            └──(PROVIDES) dep2_comp
+
+    what are the expectations in terms of sources, providers and upstreams on each of these components ?
+
+    +--------------------+---------+----------+-----------+
+    |                    | sources | provides | upstreams |
+    +--------------------+---------+----------+-----------+
+    | root_component     | 0       | 2        | 1         |
+    +--------------------+---------+----------+-----------+
+    | upstream_component | 0       | 0        | 0         |
+    +--------------------+---------+----------+-----------+
+    | dep1_component     | 1       | 1        | 0         |
+    +--------------------+---------+----------+-----------+
+    | dep2_component     | 2       | 0        | 0         |
+    +--------------------+---------+----------+-----------+
+
+    Are the following statements true ?
+
+    * upstream has 0 sources as it is its own SOURCE
+    * should dep1 or dep2 dynamically resolve the top level component's upstreams ?
+    * should dep2 list 2 sources ?
+    * we should not list upstreams in provides right ?
+
+    """
+
+    # create a top level root source component
+    root_comp = ComponentFactory(
+        name="root_comp",
+        type=Component.Type.RPM,
+        arch="src",
+        namespace=Component.Namespace.REDHAT,
+        related_url="https://example.org/related",
+    )
+    root_node, _ = ComponentNode.objects.get_or_create(
+        type=ComponentNode.ComponentNodeType.SOURCE,
+        parent=None,
+        purl=root_comp.purl,
+        defaults={"obj": root_comp},
+    )
+
+    # create upstream child node
+    upstream_comp = ComponentFactory(
+        name="upstream_comp",
+        type=Component.Type.PYPI,
+        namespace=Component.Namespace.UPSTREAM,
+        related_url="https://example.org/related",
+    )
+    ComponentNode.objects.get_or_create(
+        type=ComponentNode.ComponentNodeType.SOURCE,
+        parent=root_node,
+        purl=upstream_comp.purl,
+        defaults={"obj": upstream_comp},
+    )
+
+    # create dep component
+    dep_comp = ComponentFactory(
+        name="cool_dep_component",
+        arch="src",
```

Comment:
For a root component that is a source RPM, the child components are all binary RPMs, where the arch value is something like "aarch64" or "x86_64".
```suggestion
        arch="aarch64"
```

Reply:
done